File: md_lj_llle_module.f90
! md_lj_llle_module.f90
! Force routine for MD, LJ, Lees-Edwards, using link-lists
MODULE md_module

  !------------------------------------------------------------------------------------------------!
  ! This software was written in 2016/17                                                           !
  ! by Michael P. Allen <m.p.allen@warwick.ac.uk>/<m.p.allen@bristol.ac.uk>                        !
  ! and Dominic J. Tildesley <d.tildesley7@gmail.com> ("the authors"),                             !
  ! to accompany the book "Computer Simulation of Liquids", second edition, 2017 ("the text"),     !
  ! published by Oxford University Press ("the publishers").                                       !
  !                                                                                                !
  ! LICENCE                                                                                        !
  ! Creative Commons CC0 Public Domain Dedication.                                                 !
  ! To the extent possible under law, the authors have dedicated all copyright and related         !
  ! and neighboring rights to this software to the PUBLIC domain worldwide.                        !
  ! This software is distributed without any warranty.                                             !
  ! You should have received a copy of the CC0 Public Domain Dedication along with this software.  !
  ! If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.                               !
  !                                                                                                !
  ! DISCLAIMER                                                                                     !
  ! The authors and publishers make no warranties about the software, and disclaim liability       !
  ! for all uses of the software, to the fullest extent permitted by applicable law.               !
  ! The authors and publishers do not recommend use of this software for any purpose.              !
  ! It is made freely available, solely to clarify points made in the text. When using or citing   !
  ! the software, you should not imply endorsement by the authors or publishers.                   !
  !------------------------------------------------------------------------------------------------!

  USE, INTRINSIC :: iso_fortran_env, ONLY : output_unit, error_unit

  IMPLICIT NONE
  PRIVATE

  ! Public routines
  PUBLIC :: introduction, conclusion, allocate_arrays, deallocate_arrays
  PUBLIC :: force

  ! Public data
  INTEGER,                              PUBLIC :: n ! Number of atoms
  REAL,    DIMENSION(:,:), ALLOCATABLE, PUBLIC :: r ! Positions (3,n)
  REAL,    DIMENSION(:,:), ALLOCATABLE, PUBLIC :: v ! Velocities (3,n)
  REAL,    DIMENSION(:,:), ALLOCATABLE, PUBLIC :: f ! Forces (3,n)

  ! Private data
  REAL, PARAMETER :: r_cut_sq = 2.0**(1.0/3.0), r_cut = SQRT(r_cut_sq) ! Minimum and cutoff of WCA LJ potential

  ! Public derived type
  TYPE, PUBLIC :: potential_type ! A composite variable for interactions comprising
     REAL    :: pot ! the potential energy
     REAL    :: vir ! the virial and
     REAL    :: lap ! the Laplacian and
     REAL    :: pyx ! the off-diagonal virial part of the pressure tensor
     LOGICAL :: ovr ! a flag indicating overlap (i.e. pot too high to use)
   CONTAINS
     PROCEDURE :: add_potential_type
     GENERIC   :: OPERATOR(+) => add_potential_type
  END TYPE potential_type

CONTAINS

  FUNCTION add_potential_type ( a, b ) RESULT (c)
    IMPLICIT NONE
    TYPE(potential_type)              :: c    ! Result is the sum of
    CLASS(potential_type), INTENT(in) :: a, b ! the two inputs
    c%pot = a%pot  +   b%pot
    c%vir = a%vir  +   b%vir
    c%pyx = a%pyx  +   b%pyx
    c%lap = a%lap  +   b%lap
    c%ovr = a%ovr .OR. b%ovr
  END FUNCTION add_potential_type

  SUBROUTINE introduction
    IMPLICIT NONE

    WRITE ( unit=output_unit, fmt='(a)' ) 'WCA shifted Lennard-Jones potential'
    WRITE ( unit=output_unit, fmt='(a)' ) 'Diameter, sigma = 1'
    WRITE ( unit=output_unit, fmt='(a)' ) 'Well depth, epsilon = 1'
    WRITE ( unit=output_unit, fmt='(a)' ) 'Lees-Edwards boundaries'

  END SUBROUTINE introduction

  SUBROUTINE conclusion
    IMPLICIT NONE

    WRITE ( unit=output_unit, fmt='(a)') 'Program ends'

  END SUBROUTINE conclusion

  SUBROUTINE allocate_arrays ( box )
    USE link_list_module, ONLY : initialize_list
    IMPLICIT NONE
    REAL, INTENT(in) :: box   ! Simulation box length

    REAL :: r_cut_box

    ALLOCATE ( r(3,n), v(3,n), f(3,n) )

    r_cut_box = r_cut / box
    IF ( r_cut_box > 0.5 ) THEN
       WRITE ( unit=error_unit, fmt='(a,f15.6)' ) 'r_cut/box too large ', r_cut_box
       STOP 'Error in allocate_arrays'
    END IF

    CALL initialize_list ( n, r_cut_box )

  END SUBROUTINE allocate_arrays

  SUBROUTINE deallocate_arrays
    USE link_list_module, ONLY : finalize_list
    IMPLICIT NONE

    DEALLOCATE ( r, v, f )
    CALL finalize_list

  END SUBROUTINE deallocate_arrays

  SUBROUTINE force ( box, strain, total )
    USE link_list_module, ONLY : make_list, sc, head, list
    IMPLICIT NONE
    REAL,                 INTENT(in)  :: box    ! Simulation box length
    REAL,                 INTENT(in)  :: strain ! Shear strain
    TYPE(potential_type), INTENT(out) :: total  ! Composite of pot, vir, lap etc

    ! total%pot is the WCA LJ potential energy for whole system
    ! total%vir is the corresponding virial
    ! total%pyx is the virial part of the yx-element of the pressure tensor
    ! total%lap is the corresponding Laplacian
    ! total%ovr is a warning flag that there is an overlap
    ! This routine also calculates forces and stores them in the array f
    ! If total%ovr is set to .true., the forces etc should not be used

    ! It is assumed that positions are in units where box = 1
    ! Forces are calculated in units where sigma = 1 and epsilon = 1
    ! Uses link lists
    ! Lees-Edwards boundaries, in sliding brick arrangement
    ! Flow/gradient/vorticity directions are x/y/z == 1/2/3

    INTEGER               :: i, j, ci1, ci2, ci3, k, k_max, shift
    INTEGER, DIMENSION(3) :: ci, cj
    REAL                  :: r_cut_box, r_cut_box_sq, box_sq, rij_sq
    REAL                  :: sr2, sr6, sr12
    REAL,    DIMENSION(3) :: rij, fij
    REAL,    PARAMETER    :: sr2_ovr = 1.77 ! Overlap threshold (pot > 100)
    TYPE(potential_type)  :: pair

    ! Set up vectors to half the cells in neighbourhood of 3x3x3 cells in cubic lattice
    ! The cells are chosen so that if (d1,d2,d3) appears, then (-d1,-d2,-d3) does not.
    ! For sheared boundaries, with the chosen axes, it is most convenient to make this set
    ! include all the cells with d2=+1 and none of the ones with d2=-1
    INTEGER, PARAMETER :: nk = 13, nk_extra = nk+3
    INTEGER, DIMENSION(3,0:nk_extra), PARAMETER :: d = RESHAPE( [ &
         &    0, 0, 0,    1, 0, 0,    1, 0, 1,   -1, 0, 1,  0, 0, 1, & ! 5 cells with d2=0
         &    1, 1,-1,    1, 1, 0,    1, 1, 1,   & ! 3 cells with d1= 1, d2=1
         &    0, 1,-1,    0, 1, 0,    0, 1, 1,   & ! 3 cells with d1= 0, d2=1
         &   -1, 1,-1,   -1, 1, 0,   -1, 1, 1,   & ! 3 cells with d1=-1, d2=1
         &   -2, 1,-1,   -2, 1, 0,   -2, 1, 1 ], & ! 3 cells with d1=-2, d2=1 (extra cells, for shear)
         &  [ 3, nk_extra+1 ] )
    INTEGER, DIMENSION(3,0:nk_extra) :: dd ! will hold sheared cell indices

    r_cut_box    = r_cut / box
    r_cut_box_sq = r_cut_box ** 2
    box_sq       = box ** 2

    ! Lees-Edwards periodic boundaries
    r(1,:) = r(1,:) - ANINT ( r(2,:) ) * strain ! Extra correction in box=1 units
    r(:,:) = r(:,:) - ANINT ( r(:,:) )          ! Standard correction in box=1 units
    CALL make_list ( n, r )

    shift = FLOOR ( strain * REAL ( sc ) ) ! Strain measured in cell lengths

    ! Initialize
    f     = 0.0
    total = potential_type ( pot=0.0, vir=0.0, pyx=0.0, lap=0.0, ovr=.FALSE. )

    ! Triple loop over cells
    DO ci1 = 0, sc-1
       DO ci2 = 0, sc-1
          DO ci3 = 0, sc-1

             ci(:) = [ ci1, ci2, ci3 ] ! 3D index of i-cell
             i = head(ci1,ci2,ci3)     ! First i-atom in cell

             ! Set up correct neighbour cell indices
             IF ( ci2 == sc-1 ) THEN                 ! i-cell is in the top layer
                k_max         = nk_extra             ! Extra cells need to be checked
                dd(:,0:k_max) = d(:,0:k_max)         ! Standard list copied
                dd(1,5:k_max) = d(1,5:k_max) - shift ! All those looking up need adjustment in the x direction
             ELSE                                    ! i-cell is not in the top layer
                k_max         = nk                   ! No extra cells need checking
                dd(:,0:k_max) = d(:,0:k_max)         ! Standard list copied
             END IF

             DO ! Begin loop over i-atoms in list
                IF ( i == 0 ) EXIT ! End of link list

                DO k = 0, k_max ! Loop over neighbouring cells

                   IF ( k == 0 ) THEN
                      j = list(i) ! First j-atom is downlist from i in current cell
                   ELSE
                      cj(:) = ci(:) + dd(:,k)         ! Neighbour j-cell 3D index
                      cj(:) = MODULO ( cj(:), sc )    ! Periodic boundary correction
                      j     = head(cj(1),cj(2),cj(3)) ! First j-atom in neighbour cell
                   END IF

                   DO ! Begin loop over j-atoms in list
                      IF ( j == 0 ) EXIT ! End of link list

                      IF ( j == i ) THEN ! This should never happen
                         WRITE ( unit=error_unit, fmt='(a,3i15)' ) 'Index error', i, j, k
                         STOP 'Impossible error in force' 
                      END IF

                      rij(:) = r(:,i) - r(:,j)                    ! Separation vector
                      rij(1) = rij(1) - ANINT ( rij(2) ) * strain ! Extra correction in box=1 units
                      rij(:) = rij(:) - ANINT ( rij(:) )          ! Periodic boundary conditions in box=1 units
                      rij_sq = SUM ( rij**2 )                     ! Squared separation

                      IF ( rij_sq < r_cut_box_sq ) THEN ! Check within cutoff

                         rij_sq   = rij_sq * box_sq ! Now in sigma=1 units
                         rij(:)   = rij(:) * box    ! Now in sigma=1 units
                         sr2      = 1.0 / rij_sq    ! (sigma/rij)**2
                         pair%ovr = sr2 > sr2_ovr   ! Overlap if too close

                         sr6      = sr2 ** 3
                         sr12     = sr6 ** 2
                         pair%pot = sr12 - sr6                    ! LJ pair potential (cut but not shifted)
                         pair%vir = pair%pot + sr12               ! LJ pair virial
                         pair%pot = pair%pot + 0.25               ! WCA LJ pair potential (cut-and-shifted)
                         pair%lap = ( 22.0*sr12 - 5.0*sr6 ) * sr2 ! LJ pair Laplacian
                         fij      = rij * pair%vir * sr2          ! LJ pair forces
                         pair%pyx = rij(2)*fij(1)                 ! Off-diagonal element of pressure tensor

                         total  = total  + pair
                         f(:,i) = f(:,i) + fij
                         f(:,j) = f(:,j) - fij

                      END IF ! End check within cutoff

                      j = list(j) ! Next j-atom
                   END DO ! End loop over j-atoms in list

                END DO ! End loop over neighbouring cells

                i = list(i) ! Next i-atom
             END DO ! End loop over i-atoms in list

          END DO
       END DO
    END DO
    ! End triple loop over cells

    ! Multiply results by numerical factors
    f         = f         * 24.0       ! 24*epsilon
    total%pot = total%pot * 4.0        ! 4*epsilon
    total%vir = total%vir * 24.0 / 3.0 ! 24*epsilon and divide virial by 3
    total%pyx = total%pyx * 24.0       ! 24*epsilon
    total%lap = total%lap * 24.0 * 2.0 ! 24*epsilon and factor 2 for ij and ji

  END SUBROUTINE force

END MODULE md_module

File: python_examples/md_lj_llle_module.py
#!/usr/bin/env python3
# md_lj_llle_module.py

#------------------------------------------------------------------------------------------------#
# This software was written in 2016/17                                                           #
# by Michael P. Allen <m.p.allen@warwick.ac.uk>/<m.p.allen@bristol.ac.uk>                        #
# and Dominic J. Tildesley <d.tildesley7@gmail.com> ("the authors"),                             #
# to accompany the book "Computer Simulation of Liquids", second edition, 2017 ("the text"),     #
# published by Oxford University Press ("the publishers").                                       #
#                                                                                                #
# LICENCE                                                                                        #
# Creative Commons CC0 Public Domain Dedication.                                                 #
# To the extent possible under law, the authors have dedicated all copyright and related         #
# and neighboring rights to this software to the PUBLIC domain worldwide.                        #
# This software is distributed without any warranty.                                             #
# You should have received a copy of the CC0 Public Domain Dedication along with this software.  #
# If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.                               #
#                                                                                                #
# DISCLAIMER                                                                                     #
# The authors and publishers make no warranties about the software, and disclaim liability       #
# for all uses of the software, to the fullest extent permitted by applicable law.               #
# The authors and publishers do not recommend use of this software for any purpose.              #
# It is made freely available, solely to clarify points made in the text. When using or citing   #
# the software, you should not imply endorsement by the authors or publishers.                   #
#------------------------------------------------------------------------------------------------#

"""Force routine for MD simulation, LJ, Lees-Edwards, using neighbour lists."""

fast = True # Change this to replace NumPy force evaluation with slower Python
r_cut_sq = 2.0**(1.0/3.0)
r_cut    = r_cut_sq**(1.0/2.0) # Minimum and cutoff of WCA LJ potential

class PotentialType:
    """A composite variable for interactions."""

    def __init__(self, pot, vir, lap, pyx, ovr):
        self.pot = pot # the potential energy cut-and-shifted at r_cut
        self.vir = vir # the virial
        self.lap = lap # the Laplacian
        self.pyx = pyx # the off-diagonal virial part of the pressure tensor
        self.ovr = ovr # a flag indicating overlap (i.e. pot too high to use)

    def __add__(self, other):
        pot = self.pot +  other.pot
        vir = self.vir +  other.vir
        lap = self.lap +  other.lap
        pyx = self.pyx +  other.pyx
        ovr = self.ovr or other.ovr

        return PotentialType(pot,vir,lap,pyx,ovr)

def introduction():
    """Prints out introductory statements at start of run."""
    
    print('WCA shifted Lennard-Jones potential')
    print('Diameter, sigma = 1')
    print('Well depth, epsilon = 1')
    print('Lees-Edwards boundaries')
    if fast:
        print('Fast NumPy force routine')
    else:
        print('Slow Python force routine')
    print('Uses neighbour lists')

def conclusion():
    """Prints out concluding statements at end of run."""

    print('Program ends')

def force ( box, strain, r ):
    """Takes in box, strain, and coordinate array, and calculates forces and potentials etc."""

    import numpy as np
    from itertools import product
    import math
    
    # It is assumed that positions are in units where box = 1
    # Forces are calculated in units where sigma = 1 and epsilon = 1
    # Lees-Edwards boundaries, in sliding brick arrangement
    # Flow/gradient/vorticity directions are x/y/z == 0/1/2
    # Uses neighbour lists

    n = r.shape[0]

    # Set up vectors to half the cells in neighbourhood of 3x3x3 cells in cubic lattice
    # The cells are chosen so that if (d0,d1,d2) appears, then (-d0,-d1,-d2) does not.
    # The last three cells are extra ones, to cope with the sheared system
    d = np.array ( [ [ 0, 0,  0], [ 1, 0, 0], [ 1, 0, 1], [-1, 0, 1], [ 0, 0, 1], # 5 cells with d1=0
                     [ 1, 1, -1], [ 1, 1, 0], [ 1, 1, 1],    # 3 cells with d0= 1, d1=1
                     [ 0, 1, -1], [ 0, 1, 0], [ 0, 1, 1],    # 3 cells with d0= 0, d1=1
                     [-1, 1, -1], [-1, 1, 0], [-1, 1, 1],    # 3 cells with d0=-1, d1=1
                     [-2, 1, -1], [-2, 1, 0], [-2, 1, 1] ] ) # 3 cells with d0=-2, d1=1

    r[:,0] = r[:,0] - np.rint(r[:,1])*strain # Extra correction in box=1 units
    r      = r      - np.rint(r)             # Ensure all atoms in periodic box
    
    sr2_ovr      = 1.77 # Overlap threshold (pot > 100)
    r_cut_box    = r_cut / box
    r_cut_box_sq = r_cut_box ** 2
    box_sq       = box ** 2

    # Initialize
    f     = np.zeros_like(r)
    total = PotentialType ( pot=0.0, vir=0.0, pyx=0.0, lap=0.0, ovr=False )

    # Calculate cell index triplets
    sc = math.floor(box/r_cut)                          # Number of cells along box edge
    c  = np.floor((r+0.5)*sc).astype(np.int_)           # N*3 array of cell indices for all atoms
    assert np.all(c>=0) and np.all(c<sc), 'Index error' # Simplistic "guard" against roundoff

    shift = math.floor(strain*sc) # Strain measured in cell lengths

    if fast:
        
        # Build list of arrays, each array holding positions of atoms in a cell
        # At the same time, define a matching set of force arrays in each cell
        # i and j number the atoms in each cell; we do not refer explicitly to indices in r
        rc, fc = [], []                        # Initially empty lists of positions and forces
        for ci in product(range(sc),repeat=3): # Triple loop over cells
            mask = np.all(c==ci,axis=1)        # Mask identifies atoms in this cell
            rc.append(r[mask,:])               # Copy atom coordinates into array, add to list
            fc.append(np.zeros_like(rc[-1]))   # Zero corresponding forces, add to list

        for ci1, rci in enumerate(rc):            # Loop over i-cells, getting all atoms in each i-cell as an array
            ci = np.unravel_index(ci1,(sc,sc,sc)) # Get i-cell triple-indices
            if rci.size==0:                       # Handle empty cell
                continue

            # Set up correct neighbour cell indices
            if ci[1]==sc-1:                # i-cell is in the top layer
                dd       = d.copy()        # Standard list copied, including extra 3 cells
                dd[5:,0] = d[5:,0] - shift # All those looking up need adjustment in the x direction
            else:                          # i-cell is not in top layer
                dd = d[:-3,:].copy()       # Last three extra cells are not needed; shift is not needed
            
            for dj in dd:                                             # Loop over neighbouring j-cells
                cj  = ci + dj                                         # Compute neighbour j-cell triple-indices
                cj1 = np.ravel_multi_index(cj,(sc,sc,sc),mode='wrap') # Convert j-cell to single-index
                rcj = rc[cj1]                                         # Get atoms in j-cell as an array
                if rcj.size==0:                                       # Handle empty cell
                    continue

                rij        = rci[:,np.newaxis,:]-rcj[np.newaxis,:,:] # Separation vectors for all i and j
                rij[:,:,0] = rij[:,:,0] - np.rint(rij[:,:,1])*strain # Extra correction in box=1 units
                rij        = rij - np.rint(rij)                      # PBCs in box=1 units
                rij_sq     = np.sum(rij**2,axis=2)                   # Squared separations
                in_range   = rij_sq < r_cut_box_sq                   # Set flags for within cutoff

                if ci1==cj1:
                    np.fill_diagonal(in_range,False) # Eliminate i==j when i-cell==j-cell
                    np.fill_diagonal(rij_sq,1.0)     # Avoid divide-by-zero below

                rij_sq = rij_sq * box_sq                         # Now in sigma=1 units
                rij    = rij * box                               # Now in sigma=1 units
                sr2    = np.where ( in_range, 1.0/rij_sq, 0.0 )  # (sigma/rij)**2, only if in range
                ovr    = sr2 > sr2_ovr                           # Overlap if too close
                sr6    = sr2 ** 3
                sr12   = sr6 ** 2
                pot    = sr12 - sr6                              # LJ potential (cut but not shifted)
                vir    = pot + sr12                              # LJ virial
                pot    = np.where ( in_range, pot+0.25, 0.0 )    # WCA LJ pair potential (cut-and-shifted)
                lap    = ( 22.0*sr12 - 5.0*sr6 ) * sr2           # LJ Laplacian
                fij    = vir * sr2                               # LJ scalar part of forces
                fij    = rij * fij[:,:,np.newaxis]               # LJ pair forces
                pyx    = rij[:,:,1]*fij[:,:,0]                   # Off-diagonal element of pressure tensor

                if ci1==cj1: # Correct for double-counting ij and ji when i-cell==j-cell
                    fij = fij / 2
                    total = total + PotentialType ( pot=np.sum(pot)/2, vir=np.sum(vir)/2, 
                                                    pyx=np.sum(pyx)/2, lap=np.sum(lap)/2, ovr=np.any(ovr) )
                else:
                    total = total + PotentialType ( pot=np.sum(pot), vir=np.sum(vir), 
                                                    pyx=np.sum(pyx), lap=np.sum(lap), ovr=np.any(ovr) )

                fc[ci1][:,:] = fc[ci1][:,:] + np.sum(fij,axis=1) # Aggregate force on atoms in i-cell
                fc[cj1][:,:] = fc[cj1][:,:] - np.sum(fij,axis=0) # Aggregate force on atoms in j-cell

        # Copy forces from list of cell arrays to main force array
        for ci in product(range(sc),repeat=3):                          # Triple loop over cells
            mask      = np.all(c==ci,axis=1)                            # Mask identifies atoms in this cell
            ci1       = np.ravel_multi_index(ci,(sc,sc,sc),mode='wrap') # Single-index
            f[mask,:] = fc[ci1]                                         # Copy atom forces from correct cell

    else:
                
        # Build list of arrays, each array holding indices of atoms in a cell
        # ki and kj are atom indices in the r array; i and j number the atoms in each cell
        k_array = np.arange(n)                 # Atom indices 0..N-1
        kc = []                                # Initially empty list of indices
        for ci in product(range(sc),repeat=3): # Triple loop over cells
            mask = np.all(c==ci,axis=1)        # Mask identifies atoms in this cell
            kc.append(k_array[mask])           # Copy atom indices into array, add to list

        for ci1, kci in enumerate(kc): # Loop over i-cells, getting atom indices as an array
            ci = np.unravel_index(ci1,(sc,sc,sc)) # Get i-cell triple-indices

            # Set up correct neighbour cell indices
            if ci[1]==sc-1: # i-cell is in the top layer
                dd       = d                # Standard list copied, including extra 3 cells
                dd[5:,0] = dd[5:,0] - shift # All those looking up need adjustment in the x direction
            else:
                dd = d[:-3,:] # Last three extra cells are not needed; shift is not needed

            for dj in dd: # Loop over neighbouring j-cells
                cj  = ci + dj                                         # Compute neighbour j-cell triple-indices
                cj1 = np.ravel_multi_index(cj,(sc,sc,sc),mode='wrap') # Convert to single-index
                kcj = kc[cj1]                                         # Get indices of atoms in j-cell as an array

                for i, ki in enumerate(kci):     # Loop over individual atoms in i-cell
                    j0 = i+1 if cj1==ci1 else 0  # Only look upwards if i-cell==j-cell
                    if j0 >= kcj.size:           # Handles (redundantly) empty j-cell and the case 
                        continue                 # where j-cell==i-cell and i is last atom

                    for kj in kcj[j0:]:                          # Loop over individual atoms in j-cell
                        rij    = r[ki,:]-r[kj,:]                 # Separation vector
                        rij[0] = rij[0] - np.rint(rij[1])*strain # Extra correction in box=1 units
                        rij    = rij - np.rint(rij)              # Periodic boundary conditions in box=1 units
                        rij_sq = np.sum(rij**2)                  # Squared separation

                        if rij_sq < r_cut_box_sq: # Check within cutoff
                            rij_sq  = rij_sq * box_sq # Now in sigma=1 units
                            rij     = rij * box       # Now in sigma=1 units
                            sr2     = 1.0 / rij_sq    # (sigma/rij)**2
                            ovr     = sr2 > sr2_ovr   # Overlap if too close
                            sr6     = sr2 ** 3
                            sr12    = sr6 ** 2
                            pot     = sr12 - sr6                    # LJ potential (cut but not shifted)
                            vir     = pot + sr12                    # LJ virial
                            pot     = pot + 0.25                    # WCA LJ potential (cut-and-shifted)
                            lap     = ( 22.0*sr12 - 5.0*sr6 ) * sr2 # LJ Laplacian
                            fij     = rij * vir * sr2               # LJ forces
                            pyx     = rij[1]*fij[0]                 # Off-diagonal element of pressure tensor
                            total   = total + PotentialType ( pot=pot, vir=vir, pyx=pyx, lap=lap, ovr=ovr )
                            f[ki,:] = f[ki,:] + fij
                            f[kj,:] = f[kj,:] - fij

    # Multiply results by numerical factors
    f         = f         * 24.0       # 24*epsilon
    total.pot = total.pot * 4.0        # 4*epsilon
    total.vir = total.vir * 24.0 / 3.0 # 24*epsilon and divide virial by 3
    total.pyx = total.pyx * 24.0       # 24*epsilon
    total.lap = total.lap * 24.0 * 2.0 # 24*epsilon and factor 2 for ij and ji
    
    return total, f

